### PR TITLE
Added the ability to add a tls cert/key to a route with the -c/-k flags.

### PR DESCRIPTION
**Pull request by jayd3e**
Can now configure an HTTPS setup via the cli.  I have confirmed that both the key/certificate get into etcd.  Going to test this out on an actual app now.

### CODE REVIEW

File: route.go [hunk 1]
```diff
@@ -10,24 +16,53 @@ import (
...
 }
 
 var routeHTTPService string
+var tlsCertPath string
+var tlsKeyPath string
 
 func init() {
 	cmdRouteAddHTTP.Flag.StringVarP(&routeHTTPService, "service", "s", "", "service name to route domain to (defaults to APPNAME-web)")
+	cmdRouteAddHTTP.Flag.StringVarP(&tlsCertPath, "tls-cert", "c", "", "path to DER encoded certificate for TLS, - for stdin")
+	cmdRouteAddHTTP.Flag.StringVarP(&tlsKeyPath, "tls-key", "k", "", "path to DER encoded private key for TLS, - for stdin")
```
Comment (titanous) [inline route.go]: `s/DER/PEM/`

File: route.go [hunk 2]
```diff
@@ -37,6 +72,57 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+
+func readKey(path string, stdin []byte) ([]byte, error) {
+	if path == "-" {
+		var derBlock *pem.Block
+		for {
+			derBlock, stdin = pem.Decode(stdin)
+			if derBlock == nil {
+				break
+			}
+			if strings.Contains(derBlock.Type, "PRIVATE KEY") {
```
Comment (titanous) [inline route.go]: Can you combine these two functions into one that's slightly more generic?

File: route.go [hunk 3]
```diff
@@ -37,6 +72,57 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+				return pem.EncodeToMemory(derBlock), nil
+			}
+		}
+		log.Fatal("No private key PEM blocks found in stdin")
```
Comment (titanous) [inline route.go]: This should be returned as an error.

File: route.go (hunk 2 above)
Comment (jayd3e) [inline route.go]: Yeah sure, maybe a readCredentials() that returns both?

Comment (titanous) [inline route.go]: I was thinking the usage would be like:

``` go
key, err := readPEM("PRIVATE KEY", path, stdin)
cert, err := readPEM("CERTIFICATE", path, stdin)
```

Comment (jayd3e) [inline route.go]: Cool, I can do that.

**Comment by jayd3e** [comment]
Cleared up everything that you mentioned.  Don't pull this until I test it on an app first though.

File: route.go [hunk 4]
```diff
@@ -10,24 +16,53 @@ import (
...
+		}
+	}
+
+	tlsCert, err := readPEM("CERTIFICATE", tlsCertPath, stdin)
+	if err != nil {
+		log.Fatal("Failed to read tlsCert")
+	}
+	tlsKey, err := readPEM("PRIVATE KEY", tlsKeyPath, stdin)
+	if err != nil {
+		log.Fatal("Failed to read tlsKey")
```
Comment (titanous) [inline route.go]: Don't use log.Fatal, return the error, it will be logged for you.

File: route.go [hunk 5]
```diff
@@ -37,6 +72,40 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+}
+
+func readFile(path string) ([]byte, error) {
+	if path == "" {
+		return nil, nil
+	}
+
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		log.Printf("Failed to open %s: %s", path, err)
```
Comment (titanous) [inline route.go]: Drop this, the error will be logged.

File: route.go [hunk 6]
```diff
@@ -37,6 +72,40 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+				break
+			}
+			if derBlock.Type == typ {
+				buffer.Write(pem.EncodeToMemory(derBlock))
+			}
+		}
+		if buffer.Len() > 0 {
+			return buffer.Bytes(), nil
+		}
+		return nil, errors.New("No certificate PEM blocks found in stdin")
```
Comment (titanous) [inline route.go]: `s/certificate //`

File: route.go [hunk 7]
```diff
@@ -37,6 +72,40 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+	if path == "-" {
+		var buffer bytes.Buffer
+		var derBlock *pem.Block
+		for {
+			derBlock, stdin = pem.Decode(stdin)
+			if derBlock == nil {
+				break
+			}
+			if derBlock.Type == typ {
+				buffer.Write(pem.EncodeToMemory(derBlock))
```
Comment (titanous) [inline route.go]:
``` go
pem.Encode(&buffer, derBlock)
```

File: route.go [hunk 8]
```diff
@@ -37,6 +72,40 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
 	return nil
 }
 
+func readPEM(typ string, path string, stdin []byte) ([]byte, error) {
+	if path == "-" {
+		var buffer bytes.Buffer
```
Comment (titanous) [inline route.go]: Rename this to `buf`.

File: route.go [hunk 9]
```diff
@@ -37,6 +72,40 @@ func runRouteAddHTTP(cmd *Command, args []string, client *controller.Client) err
...
+		var derBlock *pem.Block
```
Comment (titanous) [inline route.go]: Just call this `block`.

File: route.go (hunk 5 above)
Comment (titanous) [inline route.go]: Can just do:

``` go
return ioutil.ReadFile(path)
```

**Comment by jayd3e** [comment]
Want to check it over again for issues?

**Comment by titanous** [comment]
LGTM. Squash and I'll merge.

File: route.go [hunk 10]
```diff
@@ -10,24 +15,53 @@ import (
...
+	var stdin []byte
+	var err error
+	if tlsCertPath == "-" || tlsKeyPath == "-" {
+		stdin, err = ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("Failed to read from stdin: %s", err)
+		}
+	}
+
+	tlsCert, err := readPEM("CERTIFICATE", tlsCertPath, stdin)
```
Comment (lmars) [inline route.go]: What if `tlsCertPath` is not set?

Comment (titanous) [inline route.go]: Oops, looks like the code that handles that disappeared.

Comment (jayd3e) [inline route.go]: Good catch.  So should we require both the cert AND key if either is specified?

Comment (titanous) [inline route.go]: Yeah, feel free to error if one but not the other is set.

Comment (jayd3e) [inline route.go]: K fixed in latest commit.

File: route.go [hunk 11]
```diff
@@ -10,24 +15,62 @@ import (
...
+		tlsCert, err = readPEM("CERTIFICATE", tlsCertPath, stdin)
+		if err != nil {
+			return errors.New("Failed to read tlsCert")
+		}
+		tlsKey, err = readPEM("PRIVATE KEY", tlsKeyPath, stdin)
+		if err != nil {
+			return errors.New("Failed to read tlsKey")
+		}
+	} else {
+		if tlsCertPath != "" || tlsKeyPath != "" {
```
Comment (titanous) [inline route.go]: Can bump this up to an `else if`.

File: route.go [hunk 12]
```diff
@@ -10,24 +15,62 @@ import (
...
+			}
+		}
+
+		tlsCert, err = readPEM("CERTIFICATE", tlsCertPath, stdin)
+		if err != nil {
+			return errors.New("Failed to read tlsCert")
+		}
+		tlsKey, err = readPEM("PRIVATE KEY", tlsKeyPath, stdin)
+		if err != nil {
+			return errors.New("Failed to read tlsKey")
```
Comment (titanous) [inline route.go]: Change to `TLS key` instead of `tlsKey`, ditto for `tlsCert`.

File: route.go (hunk 11 above)
Comment (jayd3e) [inline route.go]: Oops duh XD.

**Comment by jayd3e** [comment]
Squashed.

**Comment by titanous** [comment]
Thanks!

**Pull request merged**